cmd/dashboard.go

```diff
@@ -0,0 +1,84 @@
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+	"syscall"
+
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	selector         = "name=nginx-ingress-controller"
+	ingressNamespace = "kube-system"
+)
+
+var dashboardCmd = &cobra.Command{
+	Use:   "dashboard FLAG",
+	Short: "Opens the KubeApps Dashboard",
+	Long:  "Opens the KubeApps Dashboard",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		config, err := buildOutOfClusterConfig()
+		if err != nil {
+			return err
+		}
+
+		clientset, err := kubernetes.NewForConfig(config)
+		if err != nil {
+			return err
+		}
+
+		pods, err := clientset.CoreV1().Pods(ingressNamespace).List(metav1.ListOptions{LabelSelector: selector})
+		if err != nil {
+			return err
+		}
+
+		if len(pods.Items) == 0 {
+			return errors.New("nginx ingress controller pod not found, run kubeapps up first")
+		}
+
+		podName := pods.Items[0].Name
+
+		localPort, err := cmd.Flags().GetInt("port")
+		if err != nil {
+			return err
+		}
+
+		return runPortforward(podName, localPort)
+	},
+}
+
+func runPortforward(podName string, localPort int) error {
+	cmd, err := exec.LookPath("kubectl")
+	if err != nil {
+		return err
+	}
+	args := []string{"kubectl", "--namespace", ingressNamespace, "port-forward", podName, fmt.Sprintf("%d:80", localPort)}
+
+	env := os.Environ()
+
+	openInBrowser(fmt.Sprintf("http://localhost:%d", localPort))
+	return syscall.Exec(cmd, args, env)
```

Comment:
If, rather than exec'ing `kubectl port-forward`, we run it as a child process we can wait for the port-forward to be working before opening the URL in the browser. This seemed easier than dealing with process signals though.

Reply:
I guess the answer is no but does the dashboard work if we use SVC proxy URL (`/api/v1/proxy/namespaces/ns/services/dashboard:port/`)? That way we could avoid the port forward.

In other case, how long would the portforward run? how do you kill it?

Also you don't need to directly call the `kubectl` binary, in theory it is supported in the `client-go` library:
https://github.com/kubernetes/client-go/blob/master/tools/portforward/portforward.go

In any case @ngtuna implemented it in the past (not sure if we can get rid of the `kubectl` code):
https://github.com/kubeless/kubeless/blob/166b3b0e383367f258794b36b5b6544fd920a7f2/cmd/kubeless/call.go

Reply:
(just seen the issue for updating to client-go 4.0.0 to use that)

Reply:
> I guess the answer is no but does the dashboard work if we use SVC proxy URL (/api/v1/proxy/namespaces/ns/services/dashboard:port/)? That way we could avoid the port forward.

We unfortunately can't because this doesn't work well with the ingress path-based rules.

> In other case, how long would the portforward run? how do you kill it?

This is the equivalent of `exec` in bash, so `kubectl port-forward` now owns the process. It can be killed with ctrl-C. We could manage this ourselves as a child process, but this just seemed easier.

Thanks for the pointer on the port-forward implementation in Kubeless. I see that's using the old `remotecommand` package from kubernetes/kubernetes. It's probably a good idea to avoid this dependency, but only if we're able to update kubecfg to be compatible with the newer version of apimachinery.

Reply:
Okay, I am okay to go forward with this to avoid this from being blocked.

Reply:
@prydonius It's still unclear on how to do `port-forward` with client-go, even with v4.0.0. See this https://github.com/kubernetes/client-go/issues/51

WDYT ?

Reply:
@ngtuna see the links I put in https://github.com/kubeapps/kubeapps/issues/40, that issue is linking to an older version of kubectl